Formula/clickhouse.rb

```diff
@@ -0,0 +1,74 @@
+class Clickhouse < Formula
+  desc "Free analytics DBMS for big data with SQL interface"
+  homepage "https://clickhouse.tech"
+  url "https://github.com/ClickHouse/ClickHouse/releases/download/v21.6.3.14-stable/ClickHouse_sources_with_submodules.tar.gz"
```

Comment:
This seems to vendor a whole lot of dependencies. We should avoid that and use a Homebrew/core dependency whenever one is available.

See https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae.

Reply:
Thank you for your fast review.
Actually we have very thorough CI process in ClickHouse and we test master and stable versions that occur as packages and binaries for our users. You can read more about CI here https://clickhouse.tech/docs/en/development/continuous-integration/
We do not use dynamic libraries and almost even minimize glibc dependences for maximum portability.
As ClickHouse is written in C++ there is no easy way for dependencies check and updates like in Java/etc. Our community and team updates libraries and this process usually needs. 
If we start to support separate build process for brew it will be complex as lots of libraries do not release official updates(for example Poco) and we have to add bugfixes in forked repos. All changes are published in https://github.com/ClickHouse-Extras. Moreover we cannot test such releases with unbundled versions of libraries the same way as we test upstream and stable releases.
Separate build for brew means we provide our users not the same ClickHouse that is available in official builds. I understood https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae as it is better to build everything in your envionment and not just reuse build artifacts from our CI process. As you can see we do not use any system libraries and just need working compiler with c++17 support.

I see it is not an easy question and we should find solution that will benefits our communities.

Reply:
From that same page: https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Reply:
You might have your own testing infrastructure for your dependencies, but I think your coverage is also a bit limited here.

For example, it's clear that you don't test building FlatBuffers on macOS Mojave since the Mojave build fails at building FlatBuffers. If you used Homebrew FlatBuffers instead, this would not be a problem.

Reply:
Actually there is a lot of work done to make binaries which are portable. For example we check that you can use the exact same binary on centos:5/ubuntu:12.04 with really old glibc 2.4. We have cross-compile builds for aarch64, freebsd and macOS. Currently we do not have macOS servers to be able to run all necessary checks for that platform(and for different macOS versions).
We do have precompiled versions(https://clickhouse.tech/docs/en/getting-started/install/#from-binaries-non-linux) - it should be trivial to check that it works on Mojave. Probably we will get Apple servers in future and add them to CI.

FlatBuffers here is not a single problem - actually I'm almost sure it would successfully build if I add llvm dependency and build it without AppleClang on Mojave. When I encountered the same error on Catalina, I found that Xcode developer tools allow to complile project and CLT fails. Build environment should not be somehow related to functionality and we allow to use c++17 and c++20 features if it is justifiable. So, yes - we do not test _building_ on old platforms, but support running ClickHouse on older operating systems.

Imagine we will get it somehow working. Rewrite all cmake files. Add 75 dependencies to libraries that are present in Homebrew or those that are not. Export all patches from 60+ libraries that we had to write to fix bugs in those libraries. 
- We can't control that all 75 libraries have mainatainers that update brew formula for them.
- I can't understand how we can patch library source code and not only our sources. For example we have 85 commits in Poco, that was not yet taken to upstream(example https://github.com/pocoproject/poco/pull/2961)
- In normal way we have just make security fix to upstream(probably just checkout newer submodule version), test this build and switch version in brew to more recent release. Here I don't understand who exactly updates library. You force dynamic linkage or you want to rebuild all packages that depend on library in build stage?
- We know several use cases when using other versions of dynamic libraries lead to problems in stability. If we do not know exactly what versions will be user by a user we cannot check this environment. And considering we have 75 submodules, it is almost unthinkable.

And to understand what is the reason of this policy. Is it just for security reason only or there are some other ideas?

Reply:
> Actually there is a lot of work done to make binaries which are portable.

Sure; I don't doubt that. That doesn't mean your testing coverage isn't finite.

> FlatBuffers here is not a single problem - actually I'm almost sure it would successfully build if I add llvm dependency and build it without AppleClang on Mojave.

I've done a bit of work on using LLVM Clang in place of Apple Clang in order to build a formula on Mojave (see, e.g., `grpc`, `mavsdk`, `bear`, `folly`, etc), and I'm almost sure it would **not** build successfully just by using LLVM Clang in place of Apple Clang.

An unguarded `aligned_alloc` call just isn't something you can fix by using a newer compiler, because it likely involves making system calls that just aren't implemented.

> Export all patches from 60+ libraries that we had to write to fix bugs in those libraries.

You should be upstreaming these patches anyway, regardless of whether you support building ClickHouse with system-provided libraries or not.

> * We can't control that all 75 libraries have mainatainers that update brew formula for them.

Sure. This is true about any software that uses an external dependency. I would argue that attempting to exert control over external dependencies isn't really something you should be trying to do anyway.

> * I can't understand how we can patch library source code and not only our sources. For example we have 85 commits in Poco, that was not yet taken to upstream(example [pocoproject/poco#2961](https://github.com/pocoproject/poco/pull/2961))

Why haven't these patches been upstreamed?

> * In normal way we have just make security fix to upstream(probably just checkout newer submodule version), test this build and switch version in brew to more recent release. Here I don't understand who exactly updates library. You force dynamic linkage or you want to rebuild all packages that depend on library in build stage?

We link libraries dynamically.

> * We know several use cases when using other versions of dynamic libraries lead to problems in stability. If we do not know exactly what versions will be user by a user we cannot check this environment. And considering we have 75 submodules, it is almost unthinkable.

What problems are these exactly?

> And to understand what is the reason of this policy. Is it just for security reason only or there are some other ideas?

Security is one reason. De-duplication is another. It kinda defeats a big part of the use of a package manager if a package's build system includes its own specialised package manager for its vendored dependencies.